python/metatensor-learn/metatensor/learn/nn/module_map.py

```diff
@@ -0,0 +1,470 @@
+        ...                 [0, 0],
+        ...                 [0, 1],
+        ...             ]
+        ...         ),
+        ...     ),
+        ...     components=[],
+        ...     properties=Labels(["properties"], np.array([[0], [1], [2]])),
+        ... )
+        >>> block_2 = TensorBlock(
+        ...     values=torch.tensor(
+        ...         [
+        ...             [5.0, 8.0, 2.0],
+        ...             [1.0, 2.0, 8.0],
+        ...         ]
+        ...     ),
+        ...     samples=Labels(
+        ...         ["structure", "center"],
+        ...         np.array(
+        ...             [
+        ...                 [0, 0],
+        ...                 [0, 1],
+        ...             ]
+        ...         ),
+        ...     ),
+        ...     components=[],
+        ...     properties=Labels(["properties"], np.array([[3], [4], [5]])),
+        ... )
+        >>> keys = Labels(names=["key"], values=np.array([[0], [1]]))
+        >>> tensor = TensorMap(keys, [block_1, block_2])
+
+        Create modules
+
+        >>> linear = torch.nn.Linear(3, 1, bias=False)
+        >>> with torch.no_grad():
+        ...     _ = linear.weight.copy_(torch.tensor([1.0, 1.0, 1.0]))
+        ...
+        >>> modules = [linear, deepcopy(linear)]
+        >>> # you could also extend the module by some nonlinear activation function
+
+        Create ModuleMap from this ModucDict and apply it
+
+        >>> module_map = ModuleMap(tensor.keys, modules)
+        >>> out = module_map(tensor)
+        >>> out
+        TensorMap with 2 blocks
+        keys: key
+               0
+               1
+        >>> out[0].values
+        tensor([[ 7.],
+                [14.]], grad_fn=<MmBackward0>)
+        >>> out[1].values
+        tensor([[15.],
+                [11.]], grad_fn=<MmBackward0>)
+
+        Lets look at the metadata
+
+        >>> tensor[0]
+        TensorBlock
+            samples (2): ['structure', 'center']
+            components (): []
+            properties (3): ['properties']
+            gradients: None
+        >>> out[0]
+        TensorBlock
+            samples (2): ['structure', 'center']
+            components (): []
+            properties (1): ['_']
+            gradients: None
+
+        It got completely lost because we cannot know in general what the output is.
+        You can add in the initialization of the ModuleMap a TensorMap that contains
+        the intended output Labels.
+    """
+
+    def __init__(
+        self,
+        in_keys: Labels,
+        modules: Sequence[Module],
+        out_tensor: Optional[TensorMap] = None,
+    ):
+        super().__init__()
+        if len(in_keys) != len(modules):
+            raise ValueError(
+                "in_keys and modules must match in length, but found "
+                f"{len(in_keys) != len(modules)} [len(in_keys) != len(modules)]"
+            )
+        self._in_keys: Labels = in_keys
+        self._modules_list = ModuleList(modules)
```

Comment:
This is a rookie question, but why can't we just used the `ModuleDict` with this fancy interface to store the modules?

Reply:
What would be the key type? a string is not ideal since it requires creating a new string for each lookup. Here the list index can be found with a proper hash lookup (though `Labels.position(...)`), making this behave a lot like an actual dict for runtime-performance

Reply:
> This is a rookie question, but why can't we just used the ModuleDict with this fancy ModuleMapInterface to store the modules?

This is a very valid question. We did this in an earlier implementation (that can be found [here](https://github.com/lab-cosmo/equisolve/blob/c858bedef4b2799eb445e4c92535ee387224089a/src/equisolve/nn/module_tensor.py#L123C61-L123C71)). As Guilaume explained, since we cannot use `LabelEntriy`s as keys because they are not hashable (and we cannot make them hashable without dropping support for torch versions before 2.0 if I remember correctly), we use this solution of two lists in the same order.

Reply:
> and we cannot make them hashable without dropping support for torch versions before 2.0 if I remember correctly

No, we cannot make them hashable at all. `__hash__` is not in the list of allowed magic methods: https://github.com/pytorch/pytorch/blob/26956980c6953b42e57ad889b87444c4abccec1c/torch/csrc/jit/python/script_init.cpp#L652 

We could add it there, but we won't be able to use it until it reaches a stable release.